GSA-TTS/FAC
# Django FSM migration to Viewflow

Part one of #4259

This branch migrates the use of the deprecating `django-fsm` library to using `django-viewflow` instead.

- Added new `django-viewflow` and `django-filter` dependencies to `requirements.txt`.
- New file `viewflow.py` under `/audit/models/` which contains the FSM logic for transitioning an SAC.
- Moved `STATUS` enumeration outside of the `SingleAuditChecklist` class. This required import/reference changes across many files and tests.
- Removed references of old deprecated library `django-fsm`.
- New migration to handle the changing the SAC's `submission_name` field to remove dependency on the deprecated `django-fsm`.

This should be blocked until #4292 (the race condition change) is implemented, since that PR includes some potential git conflicts with this branch that will need to be resolved before merging into `main`.

**IMPORTANT** - `django-fsm` is still referenced as an import for a historical migration in `/backend/audit/`. Since we are trying to remove the `django-fsm` library, this will need to be addressed in a later PR. While this deprecated library is no longer being used anywhere in the system, it still needs to be fixed in the old migration before it can be removed from `requirements.txt`.

## How to test
First, make sure that your local is able to install the new `viewflow` dependency with `docker-clean` and `docker-first-run`.

Ensure both backend tests with `python manage.py test` and the E2E tests succeed. For bonus points, run through the submission process manually to verify that each of the states are rendering the proper name for the submission as you go through the steps.

## PR Checklist: Submitter


## PR Checklist: Reviewer


The larger the PR, the stricter we should be about these points.

## Pre Merge Checklist: Merger
